pkg/reconciler/v1alpha1/autoscaling/kpa_scaler.go

```diff
@@ -128,8 +128,22 @@ func (ks *kpaScaler) Scale(ctx context.Context, kpa *kpa.PodAutoscaler, desiredS
 	currentScale := scl.Spec.Replicas
 
 	if desiredScale == 0 {
-		// Scale to zero grace period.
-		if !kpa.Status.CanScaleToZero(ks.getAutoscalerConfig().ScaleToZeroGracePeriod) {
+		// We should only scale to zero when both of the following conditions are true:
+		//   a) The KPA has been active for atleast the idle period, after which it gets marked inactive
+		//   b) The KPA has been inactive for atleast the grace period
+		config := ks.getAutoscalerConfig()
+		ok, err := kpa.Status.CanScaleToZero(config.ScaleToZeroIdlePeriod, config.ScaleToZeroGracePeriod)
+
+		// KPA is active, but not longer than the idle period
+		if err != nil {
+			// Return error so that the KPA isn't marked inactive
+			logger.Errorf("Cannot scale to zero.", zap.Error(err))
```

Comment:
Should we really log an error in this case? From my understanding this is triggered when we try to scale to zero before activation has properly completed? If we consider that an error, should we find other means that make this not an error (for example: not proposing scale to 0 from the autoscaler if no pod metrics have arrived yet)?

Reply:
Yeah, I agree, the `return err` feels out of place but I can't think of a simpler way around the situation.

It's actually not the scale-to-zero before activation case, that's covered earlier in [`reconcile`](https://github.com/tanzeeb/serving/blob/fcb36cfb12b4ca24bc6914f683d2db5231cbfe49/pkg/reconciler/v1alpha1/autoscaling/autoscaling.go#L188-L193): 

```go
	want := metric.DesiredScale

	// Don't scale-to-zero if the KPA is still activating
	if want == 0 && kpa.Status.IsActivating() {
		want = -1
	}
```

This is actually the case where the revision has successfully activated, but is not receiving any traffic. Since the autoscaler metrics have no knowledge of pod readiness (only the reconciler does), the `desiredScale` is 0 even if the revision has only been active for 1 second. As it stands, the reconciler can't tell the difference between an inactive target that has been around for a long time and an inactive target that has just become active.

The target should be allowed to be active for atleast the idle period, but if we don't return an error, the [reconciler will mark it inactive](https://github.com/tanzeeb/serving/blob/fcb36cfb12b4ca24bc6914f683d2db5231cbfe49/pkg/reconciler/v1alpha1/autoscaling/autoscaling.go#L221-L231).

Reply:
Some other approaches to this that I can think of:

* move this logic to the reconciler, but then the reconciler will also need its own [autoscaler config](https://github.com/tanzeeb/serving/blob/fcb36cfb12b4ca24bc6914f683d2db5231cbfe49/pkg/reconciler/v1alpha1/autoscaling/kpa_scaler.go#L65-L85)

* change the `Scale` function signature to return an "actual desired scale", which could signal to the reconciler the difference in opinion between the metrics and the KPA scaler (ie, in this case it would be called with `desiredScale = 0`, but would return `actualDesiredScale = 1`)

Happy to try the second approach if there's a strong objection to this `return err` in the KPA scaler?

Reply:
@tanzeeb is the second approach basically the same thing proposed [here](https://github.com/knative/serving/pull/2244#discussion_r229767281) and [here](https://github.com/knative/serving/pull/2378#discussion_r230153673). If yes, that sounds good to me.

Reply:
Yes, exactly, in fact the proposed signature in the second PR is even nicer!

Reply:
Should we just wait for #2244 and #2378 to be merged?

Reply:
Nah, please go ahead. We can always rebase.

Reply:
@josephburnett yup, I'll take care of it

Reply:
Looks like #2378 was merged with the signature change, rebased.